File: Python/Word Search.py
class Solution:
# @param board, a list of lists of 1 length string
# @param word, a string
# @return a boolean
    def exist(self, board, word):
        visited = [[False for m in range(len(board[0]))] for n in range(len(board))] #Key improvement
        for i in range(0,len(board)):
            for j in range(0,len(board[0])):
                if board[i][j] == word[0]:
                    if self.DFS(board, word, visited, 0, i, j):
                        return True
        return False

    def DFS(self,board, word, visited, index, i, j):
        if i >= 0 and j >=0 and i < len(board) and j < len(board[0]) and board[i][j] == word[index] and visited[i][j] == False:
            visited[i][j] = True
            if index + 1 == len(word):
                return True
            if self.DFS(board, word, visited, index + 1, i + 1, j):
                return True
            if self.DFS(board, word, visited, index + 1, i - 1, j):
                return True
            if self.DFS(board, word, visited, index + 1, i , j + 1):
                return True
            if self.DFS(board, word, visited, index + 1, i , j - 1):
                return True
            visited[i][j]=False #Key improvement
        return False

File: Java/src/Word Search.java
import java.util.Arrays;

public class Solution
{
  public boolean exist(char[][] board, String word)
  {
    char[] charWord = word.toCharArray();
    boolean[][] visited = new boolean[board.length][board[0].length];
    for (int x = 0; x < visited.length; x++)
    {
      Arrays.fill(visited[x], false);
    }
    for (int i = 0; i < board.length; i++)
    {
      for (int j = 0; j < board[0].length; j++)
      {
        if (board[i][j] == charWord[0])
        {
          if (DFS(board, i, j, charWord, 0, visited))
            return true;
        }
      }
    }
    return false;
  }

  private boolean DFS(char[][] board, int i, int j, char[] charWord, int index, boolean[][] visited)
  {
    if (index == charWord.length)
      return true;
    if (j < 0 || i < 0 || i == board.length || j == board[0].length || visited[i][j])
      return false;
    if (board[i][j] != charWord[index])
      return false;

    visited[i][j] = true;
    if (DFS(board, i + 1, j, charWord, index + 1, visited))
      return true;
    if (DFS(board, i - 1, j, charWord, index + 1, visited))
      return true;
    if (DFS(board, i, j - 1, charWord, index + 1, visited))
      return true;
    if (DFS(board, i, j + 1, charWord, index + 1, visited))
      return true;
    visited[i][j] = false;
    return false;
  }
}
